services/QuillLMS/spec/controllers/admins_controller_spec.rb

```diff
@@ -251,6 +251,141 @@
+        post :deny_teacher_verification, params: { id: teacher.id }
+        teacher.reload
+        expect(teacher.teacher_verifier_id).to eq(admin.id)
+        expect(teacher.teacher_verification_status).to eq(TeacherInfo::DENIED)
+      end
+
+      it 'enqueues the analytics and email workers for denial' do
+        expect(TeacherVerificationDeniedAnalyticsWorker).to receive(:perform_async).with(teacher.id)
+        expect(TeacherVerificationDeniedEmailWorker).to receive(:perform_async).with(teacher.id)
+        post :deny_teacher_verification, params: { id: teacher.id }
+      end
+    end
+
+    context 'when teacher verification request is already processed' do
+      before do
+        teacher.teacher_info.update(verifier_id: 1, verification_status: TeacherInfo::APPROVED)
+      end
+
+      it 'redirects to the teacher verification request already processed page' do
+        post :deny_teacher_verification, params: { id: teacher.id }
+        expect(response).to redirect_to("/users/#{teacher.id}/teacher_verification_request_already_processed")
+      end
+
+      it 'does not update the teacher record' do
+        original_status = teacher.teacher_verification_status
+        post :deny_teacher_verification, params: { id: teacher.id }
+        teacher.reload
+        expect(teacher.teacher_verification_status).to eq(original_status)
+      end
+    end
+  end
+
+  describe '#teacher_verification_request_already_processed' do
+    it 'renders successfully' do
+      get :teacher_verification_request_already_processed, params: { id: teacher.id }
+      expect(response).to be_successful
+    end
+  end
+
+  describe '#approve_teacher_verifications' do
+    let!(:teacher2) { create(:teacher_with_school) }
+
+    it 'updates each teacher record to approved and sets the current teacher verifier' do
+      post :approve_teacher_verifications, params: { teacher_ids: [teacher.id, teacher2.id] }
+
+      teacher.reload
+      teacher2.reload
+
+      expect(teacher.teacher_verifier_id).to eq(admin.id)
+      expect(teacher.teacher_verification_status).to eq(TeacherInfo::APPROVED)
+      expect(teacher2.teacher_verifier_id).to eq(admin.id)
+      expect(teacher2.teacher_verification_status).to eq(TeacherInfo::APPROVED)
+    end
+
+    it 'enqueues analytics and email jobs for each teacher' do
+      expect {
+        post :approve_teacher_verifications, params: { teacher_ids: [teacher.id, teacher2.id] }
+      }.to change(TeacherVerificationApprovedAnalyticsWorker.jobs, :size).by(2)
+        .and change(TeacherVerificationApprovedEmailWorker.jobs, :size).by(2)
+    end
+
+    it 'returns a pluralized success message' do
+      post :approve_teacher_verifications, params: { teacher_ids: [teacher.id, teacher2.id] }
+      response_message = JSON.parse(response.body)['message']
+      expect(response_message).to eq(I18n.t('admin.approve_teacher_verifications'))
```

Comment:
Take or leave this, but having the constant string defined here makes me wonder if the controller should actually define these as constants directly, so that you could references those constants in the spec:
```ruby
class AdminsController
  APPROVE_TEACHER_VERIFICATIONS_MESSAGE = 'admin.approve_teacher_verifications'
end
```
```
expect(response_message).to eq(described_class::APPROVE_TEACHER_VERIFICATIONS_MESSAGE)
```